tv_apps/tv-deck/js/tv_deck.js

```diff
@@ -0,0 +1,538 @@
+          return;
+        }
+        clearTimeout(this.panelTimeoutId);
+        this.panelTimeoutId =
+            setTimeout(this._onPanelTimeout.bind(this), this.panelTimeoutDelay);
+        return;
+      default:
+        return;
+    }
+
+    channelIndex = (channelIndex < 0) ? channelList.length - 1 : channelIndex;
+    channelIndex = (channelIndex > channelList.length - 1) ? 0 : channelIndex;
+
+    // If a channel is not found, add flash animation.
+    if (!channelList[channelIndex]) {
+      this.channelPanel.classList.add('flash');
+      return;
+    }
+
+    this.playingChannelNumber = channelList[channelIndex].channel.number;
+    this.setHash();
+  };
+
+  /**
+   * onEnterNumber handles event when an user enters a number through keyboard.
+   * Channel will be switched after enterNumberTimeoutDelay.
+   */
+  proto._onEnterNumber = function td__onEnterNumber(evt) {
+    var number = evt.keyCode - 48;
+    if (number >= 0 && number <= 9) {
+      if (!this.enterNumberTimeoutId) {
+        this.channelNumber.textContent = '';
+      }
+
+      clearTimeout(this.enterNumberTimeoutId);
+      clearTimeout(this.panelTimeoutId);
+      this.enterNumberTimeoutId =
+            setTimeout(this._onEnter.bind(this), this.enterNumberTimeoutDelay);
+      this.channelPanel.classList.remove('hidden');
+
+      // length is limit to 4 digits
+      if (this.channelNumber.textContent.length < 4) {
+        this.channelNumber.textContent += number;
+      }
+    }
+  };
+
+  /**
+   * onEnter switch channels accroding to channelNumber element. It will be
+   * triggerred either by pressing enter key or enterNumebrTimeout.
+   */
+  proto._onEnter = function td__onEnter() {
+    if (this.enterNumberTimeoutId) {
+      clearTimeout(this.enterNumberTimeoutId);
+      this.enterNumberTimeoutId = null;
+
+      clearTimeout(this.panelTimeoutId);
+      this.panelTimeoutId =
+            setTimeout(this._onPanelTimeout.bind(this), this.panelTimeoutDelay);
+
+      var newNumber = this.channelNumber.textContent;
+
+      // reset back to current channel number if the number entered
+      // is not valid.
+      if (this.playingSource.channelIndexHash[newNumber] !== 0 &&
+          !this.playingSource.channelIndexHash[newNumber]) {
+        this.channelNumber.textContent = this.playingChannelNumber;
+        this.channelPanel.classList.add('flash');
+        return;
+      }
+
+      this.playingChannelNumber = newNumber;
+      this.setHash();
+    }
+  };
+
+  proto._showErrorState = function td__showErrorState() {
+    this.updateChannelInfo('---', '--');
+    this.tvStreamElement.src = null;
+  };
+
+  proto._rotateLoadingIcon = function td__rotateLoadingIcon() {
+    if (!this.loadingIcon) {
+      return;
+    }
+    this.loadingIcon.classList.remove('loading-circle');
+    this.loadingIcon.classList.remove('hidden');
+    window.getComputedStyle(this.loadingIcon).width;
+    window.getComputedStyle(this.loadingIcon, ':before').width;
```

Comment:
Do we really need to call ":before" and ":after"? It should be calculated by once getComputedStyle.